parlai/mturk/core/worker_manager.py

```diff
@@ -0,0 +1,344 @@
+import parlai.mturk.core.shared_utils as shared_utils
+
+
+# Time to persist a disconnect before forgetting about it. Combined with the
+# above this will block workers that disconnect at least 10 times in a day
+DISCONNECT_PERSIST_LENGTH = 60 * 60 * 24
+
+# Max number of conversation disconnects before a turker should be blocked
+MAX_DISCONNECTS = 5
+
+DISCONNECT_FILE_NAME = 'disconnects.pickle'
+
+parent_dir = os.path.dirname(os.path.abspath(__file__))
+
+
+class WorkerState():
+    """Class for holding state information about an mturk worker"""
+    def __init__(self, worker_id, disconnects=0):
+        """Create a new worker state for the given worker_id. Number of
+        prior disconnects is optional.
+        """
+        self.worker_id = worker_id
+        self.agents = {}
+        self.disconnects = disconnects
+
+    def active_conversation_count(self):
+        """Return the number of conversations within this worker state
+        that aren't in a final state
+        """
+        count = 0
+        for assign_id in self.agents:
+            if not self.agents[assign_id].is_final():
+                count += 1
+        return count
+
+    def add_agent(self, mturk_agent):
+        """Add an assignment to this worker state with the given assign_it"""
+        assert mturk_agent.worker_id == self.worker_id, \
+            "Can't add agent that does not match state's worker_id"
+        self.agents[mturk_agent.assignment_id] = mturk_agent
+
+    def get_agent_for_assignment(self, assignment_id):
+        return self.agents.get(assignment_id, None)
+
+    def has_assignment(self, assign_id):
+        """Returns true if this worker has an assignment for the given id"""
+        return assign_id in self.agents
+
+    def completed_assignments(self):
+        """Returns the number of assignments this worker has completed"""
+        complete_count = 0
+        for agent in self.agents.values():
+            if agent.get_status() == AssignState.STATUS_DONE:
+                complete_count += 1
+        return complete_count
+
+
+class WorkerManager():
+    """Class used to keep track of workers state, as well as processing
+    messages that come from the web client.
+    """
+
+    def __init__(self, mturk_manager, opt):
+        self.opt = opt
+        self.mturk_manager = mturk_manager
+        self.mturk_workers = {}
+        self.conv_to_agent = {}
+        self.assignment_to_worker_id = {}
+        self.hit_id_to_agent = {}
+        self.time_blocked_workers = []
+        self.load_disconnects()
+        self.is_sandbox = mturk_manager.is_sandbox
+
+    def _create_agent(self, hit_id, assignment_id, worker_id):
+        """Initialize an agent and return it"""
+        return MTurkAgent(
+            self.opt, self.mturk_manager, hit_id, assignment_id, worker_id)
+
+    def _get_worker(self, worker_id):
+        """A safe way to get a worker by worker_id"""
+        if worker_id in self.mturk_workers:
```

Comment:
why not e.g. `self.mturk_workers.get(worker_id, None)`

Reply:
copypasta - yours is cleaner